explorer/interpreter/action_stack.cpp

```diff
@@ -155,16 +176,48 @@ auto ActionStack::FinishAction(Nonnull<const Value*> result)
     case Action::Kind::ExpressionAction:
     case Action::Kind::LValAction:
     case Action::Kind::PatternAction:
-      PopScopes();
+      destroy_actions_ = PopScopes();
       SetResult(result);
+      break;
+    case Action::Kind::CleanUpAction:
+      break;
+  }
+  if (!destroy_actions_.empty()) {
+    for (auto& x : destroy_actions_) {
+      auto& scope = x->scope();
+      if (scope && x->kind() != Action::Kind::CleanUpAction) {
+        std::unique_ptr<Action> cleanup_action =
+            std::make_unique<CleanupAction>(std::move(*scope));
+        todo_.Push(std::move(cleanup_action));
+      }
+    }
+    destroy_actions_.clear();
+  }
+  auto& scope = act->scope();
+  if (scope && act->kind() != Action::Kind::CleanUpAction) {
+    std::unique_ptr<Action> cleanup_action =
+        std::make_unique<CleanupAction>(std::move(*scope));
+    todo_.Push(std::move(cleanup_action));
   }
+
   return Success();
 }
 
 auto ActionStack::Spawn(std::unique_ptr<Action> child) -> ErrorOr<Success> {
   Action& action = *todo_.Top();
   action.set_pos(action.pos() + 1);
   todo_.Push(std::move(child));
+  if (!destroy_actions_.empty()) {
+    for (auto& x : destroy_actions_) {
+      auto& scope = x->scope();
+      if (scope && x->kind() != Action::Kind::CleanUpAction) {
+        std::unique_ptr<Action> cleanup_action =
+            std::make_unique<CleanupAction>(std::move(*scope));
+        todo_.Push(std::move(cleanup_action));
+      }
+    }
+    destroy_actions_.clear();
+  }
```

Comment:
This same block of logic seems to get repeated in this file: can you make a helper function?

Reply:
Done!